cw/util/trace_library.py

```diff
@@ -0,0 +1,221 @@
+        )
+        self.MetadataDB = db.Table(
+            "metadata",
+            self.metadata,
+            db.Column("config", db.String),
+            db.Column("datetime", db.DateTime),
+            db.Column("bitstream", db.String),
+            db.Column("binary", db.String),
+            db.Column("offset", db.Integer),
+            db.Column("sample_rate", db.Integer),
+            db.Column("scope_gain", db.Float),
+            db.Column("trigger_level", db.Float),
+            db.Column("time_div", db.Float)
+        )
+        self.metadata.create_all(self.engine)
+        self.trace_mem = []
+        self.trace_mem_ctr = 0
+        self.trace_mem_thr = trace_threshold
+        self.settingsDict = {}
+
+    def commit_traces(self):
+        """ Writes traces from memory into database.
+        """
+        if len(self.trace_mem):
+            query = db.insert(self.TracesDB)
+            traces = []
+            for trace in self.trace_mem:
+                tr = asdict(trace)
+                del tr["trace_id"]
+                traces.append(tr)
+            self.session.execute(query, traces)
+            self.session.commit()
+            self.trace_mem_ctr = 0
+            self.trace_mem = []
+
+    def write_trace(self, trace):
+        """ Write traces into database.
+
+        The trace first is written into memory. When reaching the trace
+        memory threshold, the traces are flushed into the database storage.
+
+        Args:
+            trace: The trace to write.
+        """
+        self.trace_mem.append(trace)
+        self.trace_mem_ctr += 1
+        if (self.trace_mem_ctr >= self.trace_mem_thr):
+            self.commit_traces()
+
+    def get_traces(self, start=0, end=0):
+        """ Get traces from database and stored into RAM.
+
+        Fetch traces from start to end from database storage into RAM.
+
+        Args:
+            start: fetch traces from trace index start
+            end: to trace index end. If no start and end is provided, all
+                 all traces are returned.
+        Returns:
+            The traces from the database.
+        """
+        self.commit_traces()
+
+        if end == 0:
+            query = db.select(self.TracesDB)
+        else:
+            query = db.select(self.TracesDB).where(
+                (self.TracesDB.c.trace_id >= start) &
+                (self.TracesDB.c.trace_id <= end))
+
+        ResultProxy = self.session.execute(query)
+
+        traces = []
+        for trace in ResultProxy.fetchall():
+            traces.append(Trace(**trace._mapping))
+        return traces
+
+    def get_waves(self):
+        """ Get all waves from the database.
+
+        Returns:
+            The bytearray waves from the database.
+        """
+        self.commit_traces()
+        traces = self.get_traces()
+        waves = []
+        for trace in traces:
+            waves.append(trace.wave)
+        return waves
+
+    def get_waves_int(self):
```

Comment:
More general question (putting it here because I couldn't find out where the data format is defined): What is the storage format? Is it doubles or floats? In my view it should either be int16 or when using 8-bit scopes int8. Everything else is probably very wasteful in terms of memory, disk space and compute (when converting on the fly upon reading). To be honest, I think int8 could even be skipped in favor of just int16 to simplify things.

Reply:
Thanks for explaining, I wasn't aware that we don't store numpy files. Follow-up question: is everything in one file? What happens if you you've stored 100k traces to disk already and you want to append the next 100k traces, will it store all 200k traces again (because it's just one file)?

Reply:
The data is stored in a single SQL database file.
When using `write_to_buffer`, traces are stored in memory. With `flush_to_disk`, this memory buffer is committed into the database file. If there are already 100k traces stored and you want to flush another 100k traces from memory into the database file,  `flush_to_disk` just appends these 100k traces to the existing 100k.

Reply:
This is perfect, thanks for explaining!

Reply:
Based on our discussion, I added fields for storing the binary and the bitstream.